File: 560.和为k的子数组.java
import java.util.HashMap;

/*
 * @lc app=leetcode.cn id=560 lang=java
 *
 * [560] 和为K的子数组
 */

// @lc code=start
class Solution {
    public int subarraySum(int[] nums, int k) {
        int[] sums = new int[nums.length+1];
        sums[0] = 0;
        int res = 0;
        for (int i = 1; i < nums.length+1; i++) {
            sums[i] = sums[i-1] + nums[i-1];
            for (int j = 0; j < i; j++) {
                if (sums[i] - sums[j] == k) {
                    res++;
                }
            }
        }
        return res;
    }

    public int subarraySum0(int[] nums, int k) {
        HashMap<Integer, Integer> map = new HashMap<Integer, Integer>();
        map.put(0, 1);
        int sum = 0;
        int res = 0;
        for (int num : nums) {
            sum += num;
            res += map.getOrDefault(sum-k, 0);
            map.put(sum, map.getOrDefault(sum, 0) + 1);
        }
        return res;
    }
}
// @lc code=end

File: 560.和为k的子数组.py
#!/usr/bin/env python
# coding=utf-8
'''
Description: 
Author: yangyuxiang
Date: 2021-05-18 09:48:11
LastEditors: Yuxiang Yang
LastEditTime: 2021-08-26 08:44:12
FilePath: /leetcode/560.和为k的子数组.py
'''
#
# @lc app=leetcode.cn id=560 lang=python
#
# [560] 和为K的子数组
#

# @lc code=start
class Solution(object):
    def subarraySum(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        sum = {0:1}
        res = 0
        sum_i = 0
        for i in range(0, len(nums)):
            sum_i = sum_i + nums[i]
            
            sum_j = sum_i - k
            if sum_j in sum:
                res += sum[sum_j]
            
            if sum_i not in sum:
                sum[sum_i] = 1
            else:
                sum[sum_i] += 1


                    
        return res

if __name__ == '__main__':
    nums = [1]
    k=0
    print(Solution().subarraySum(nums, k))
                

            
        
# @lc code=end

